File: challenge/inventory/PetInventory.hs
-- could not compile bc of stack install hell

import System.Random (newStdGen)
import System.Random.Shuffle (shuffle')
import Data.List
import Data.Text

data Pet = Pet { name :: String, says :: String }
type Dog = Pet
type Cat = Pet

toString :: Pet -> String
toString p = p.name

speak :: Pet -> String
speak p = p.says

speak :: Pet -> String
speak pet = "?"

speak :: Dog -> String
speak dog = "Woof"

speak :: Cat -> String
speak cat = "Meow"

rollCall :: [Pet] -> String
rollCall pets = intercalate "\n" messages
    where 
        messages = map greet pets
        greet pet = printf "%s says \"%s\"!" (toString pet) (speak pet)

main :: IO()
main = do

    let 
        generator = newStdGen

        dogNames = [ "Bowser", "Rex", "Professor McMann" ]
        catNames = [ "Fluffy", "Miss Thang", "Monique" ]

        dogs = map Dog dogNames :: [Dog]
        cats = map Cat catNames :: [Cat]
        pets = [dogs, cats]

        -- pets = concat 
        --         $ map (\(d, c) -> [Dog d, Cat c]) 
        --         $ zip dogs cats

        pets = pets + Pet "Baron von Fiend"

        pets = shuffle' pets (length pets) generator
        
        msg = rollCall pets

    putStrLn(msg)

File: java/PetInventory.java
import java.util.*;
import java.util.stream.*;
import java.util.function.*;

class PetInventory {

    public static void main(String[] args) {

        String[] dogNames = new String[] {
            "Bowser",
            "Rex",
            "Professor McMann"
        };

        String[] catNames = new String[] {
            "Fluffy",
            "Miss Thang",
            "Monique"
        };


        Stream<Dog> dogs = Arrays.stream(dogNames).map(Dog::new);
        Stream<Cat> cats = Arrays.stream(catNames).map(Cat::new);

        //// Another approach
        //   enum PetType { DOG, CAT };
        //
        //   Map<PetType, String[]> petInventory = Map.of(
        //           PetType.DOG, 
        //           new String[] {"Bowser", "Rex", "Professor McMann"},
        //
        //           PetType.CAT, 
        //           new String[] { "Fluffy", "Miss Thang", "Monique"});
        //
        //   Stream<Dog> dogs = Arrays.stream(petInventory.get(PetType.DOG))
        //       .map(Dog::new);
        //
        //   Stream<Cat> cats = Arrays.stream(petInventory.get(PetType.CAT))
        //       .map(Cat::new);

        List<Pet> pets = Stream.concat(dogs, cats)
            .collect(Collectors.toList());

        pets.add(new Pet("Baron von Fiend"));

        Collections.shuffle(pets);

        String petRollCall = pets.stream()
            .map(p -> String.format("%s says \"%s\"!", p, p.speak()))
            .collect(Collectors.joining("\n"));

        System.out.println(petRollCall);
    }
}


class Pet {
    private String name;
    protected String says = "?";

    public Pet(String name) {
        this.name = name;
    }

    public Pet(String name, String says) {
        this(name);
        this.says = says;
    }

    public String speak() {
        return this.says;
    }

    public String toString() {
        return this.name;
    }
}

class Dog extends Pet {
    public Dog(String name) {
        super(name, "Woof");
    }
}

class Cat extends Pet {
    public Cat(String name) {
        super(name, "Meow");
    }
}



